File: java/lib/src/test/java/com/zourzouvillys/httpsig/SecurityTest.java
package com.zourzouvillys.httpsig;

import static org.junit.jupiter.api.Assertions.*;

import java.net.URI;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;

/**
 * Security regression tests for HTTP message signatures.
 *
 * These verify that the verifier actually enforces the guardrails that
 * prevent replay attacks, algorithm confusion, and untrusted metadata leaking.
 */
class SecurityTest {

    private static final byte[] SECRET = "test-secret-key-long-enough-for-hmac!".getBytes();
    private static final String KEY_ID = "test-key";

    private static Keys.HmacKey hmacKey() {
        return Keys.hmacSHA256Key(KEY_ID, SECRET);
    }

    /**
     * Helper: sign a minimal GET request with the given params, return the signed message.
     */
    private static HttpMessage signRequest(SignatureParameters params) throws HttpSigException {
        var request = RawMessage.request("GET", URI.create("https://example.com/"),
            Map.of());

        var result = Signer.sign(request, "sig1", params, hmacKey(), null);

        return RawMessage.request("GET", URI.create("https://example.com/"),
            Map.of(
                "signature-input", List.of(Signer.signatureInputHeader(result)),
                "signature", List.of(Signer.signatureHeader(result))
            ));
    }

    // -----------------------------------------------------------------------
    // 1. Future-dated signature rejection
    // -----------------------------------------------------------------------

    @Test
    void futureDatedSignature_rejectedByTightClockSkew() throws Exception {
        Instant now = Instant.now();
        Instant oneHourFromNow = now.plusSeconds(3600);

        var params = SignatureParameters.builder()
            .component("@method")
            .keyId(KEY_ID)
            .algorithm(Algorithm.HMAC_SHA256)
            .created(oneHourFromNow)
            .build();

        var signedRequest = signRequest(params);

        // 30s skew tolerance should reject a signature created 1 hour in the future
        var tightOptions = new Verifier.VerifyOptions(
            null,
            null,
            Duration.ofSeconds(30),
            null,
            null,
            () -> now
        );

        var ex = assertThrows(HttpSigException.class,
            () -> Verifier.verify(signedRequest, (keyId, alg) -> hmacKey(), tightOptions, null));
        assertTrue(ex.getMessage().contains("future-dated"),
            "expected 'future-dated' in message but got: " + ex.getMessage());
    }

    @Test
    void futureDatedSignature_acceptedByLooseClockSkew() throws Exception {
        Instant now = Instant.now();
        Instant oneHourFromNow = now.plusSeconds(3600);

        var params = SignatureParameters.builder()
            .component("@method")
            .keyId(KEY_ID)
            .algorithm(Algorithm.HMAC_SHA256)
            .created(oneHourFromNow)
            .build();

        var signedRequest = signRequest(params);

        // 2 hour tolerance should accept a signature 1 hour in the future
        var looseOptions = new Verifier.VerifyOptions(
            null,
            null,
            Duration.ofHours(2),
            null,
            null,
            () -> now
        );

        var result = Verifier.verify(signedRequest, (keyId, alg) -> hmacKey(), looseOptions, null);
        assertEquals("sig1", result.label());
    }

    @Test
    void futureDatedSignature_acceptedWhenNoClockSkewConfigured() throws Exception {
        Instant now = Instant.now();
        Instant oneHourFromNow = now.plusSeconds(3600);

        var params = SignatureParameters.builder()
            .component("@method")
            .keyId(KEY_ID)
            .algorithm(Algorithm.HMAC_SHA256)
            .created(oneHourFromNow)
            .build();

        var signedRequest = signRequest(params);

        // null maxClockSkew means no future-date check at all
        var noSkewOptions = new Verifier.VerifyOptions(
            null,
            null,
            null,
            null,
            null,
            () -> now
        );

        var result = Verifier.verify(signedRequest, (keyId, alg) -> hmacKey(), noSkewOptions, null);
        assertEquals("sig1", result.label());
    }

    // -----------------------------------------------------------------------
    // 2. Algorithm mismatch rejection
    // -----------------------------------------------------------------------

    @Test
    void algorithmMismatch_signatureInputClaimsDifferentAlgorithm() throws Exception {
        // Sign with HMAC-SHA256 and include alg param in signature
        var params = SignatureParameters.builder()
            .component("@method")
            .keyId(KEY_ID)
            .algorithm(Algorithm.HMAC_SHA256)
            .createdEpoch(1618884473L)
            .build();

        var request = RawMessage.request("GET", URI.create("https://example.com/"),
            Map.of());

        var result = Signer.sign(request, "sig1", params, hmacKey(), null);
        String sigInputHeader = Signer.signatureInputHeader(result);
        String sigHeader = Signer.signatureHeader(result);

        // Tamper: replace alg="hmac-sha256" with alg="ed25519" in the signature-input
        String tamperedSigInput = sigInputHeader.replace("hmac-sha256", "ed25519");

        // Sanity: the replacement actually happened
        assertTrue(tamperedSigInput.contains("ed25519"),
            "tampered input should contain ed25519");
        assertFalse(tamperedSigInput.contains("hmac-sha256"),
            "tampered input should not contain hmac-sha256");

        var signedRequest = RawMessage.request("GET", URI.create("https://example.com/"),
            Map.of(
                "signature-input", List.of(tamperedSigInput),
                "signature", List.of(sigHeader)
            ));

        // The key provider returns the real HMAC key. The verifier should
        // reject because the header claims ed25519 but the key is hmac-sha256.
        var ex = assertThrows(HttpSigException.class,
            () -> Verifier.verify(signedRequest, (keyId, alg) -> hmacKey(),
                Verifier.VerifyOptions.defaults(), null));
        assertTrue(ex.getMessage().contains("algorithm mismatch"),
            "expected 'algorithm mismatch' in message but got: " + ex.getMessage());
    }

    // -----------------------------------------------------------------------
    // 3. VerifyResult returns the key's algorithm, not untrusted header alg
    // -----------------------------------------------------------------------

    @Test
    void verifyResult_reportsKeyAlgorithm_notHeaderAlgorithm() throws Exception {
        var params = SignatureParameters.builder()
            .component("@method")
            .keyId(KEY_ID)
            .algorithm(Algorithm.HMAC_SHA256)
            .createdEpoch(1618884473L)
            .build();

        var signedRequest = signRequest(params);

        var result = Verifier.verify(signedRequest, (keyId, alg) -> hmacKey(),
            Verifier.VerifyOptions.defaults(), null);

        // The result's algorithm must come from the key, not from parsing the header.
        assertEquals(Algorithm.HMAC_SHA256, result.algorithm());
        assertEquals(hmacKey().algorithm(), result.algorithm());
    }

    @Test
    void verifyResult_reportsKeyAlgorithm_evenWhenHeaderOmitsAlg() throws Exception {
        // Sign WITHOUT setting algorithm in params (no alg= in signature-input)
        var params = SignatureParameters.builder()
            .component("@method")
            .keyId(KEY_ID)
            .createdEpoch(1618884473L)
            .build();

        var signedRequest = signRequest(params);

        var result = Verifier.verify(signedRequest, (keyId, alg) -> hmacKey(),
            Verifier.VerifyOptions.defaults(), null);

        // Even with no alg in the header, result.algorithm() should be the key's algorithm.
        assertEquals(Algorithm.HMAC_SHA256, result.algorithm());
    }
}

File: kotlin/lib/src/test/kotlin/io/zrz/httpsig/SecurityTest.kt
package io.zrz.httpsig

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.net.URI
import java.time.Duration
import java.time.Instant
import java.util.function.Supplier

/**
 * Security regression tests: clock skew enforcement, algorithm mismatch
 * rejection, and VerifyResult field correctness.
 */
class SecurityTest {

    companion object {
        private val SECRET = "hmac-secret-key-that-is-at-least-256-bits-long!!".toByteArray()
        private const val KEY_ID = "sec-test-key"

        private fun hmacKey(): Keys.HmacKey = Keys.hmacSHA256Key(KEY_ID, SECRET)

        private fun simpleMessage(): RawMessage.Request = RawMessage.request(
            "POST",
            URI.create("https://example.com/resource"),
            mapOf("content-type" to listOf("application/json")),
        )

        /**
         * Sign the simple message, returning a signed [RawMessage] ready for verification.
         * The [created] epoch is baked into the Signature-Input.
         */
        private fun signMessage(
            createdEpoch: Long,
            includeAlgorithm: Boolean = false,
        ): RawMessage.Request {
            val key = hmacKey()
            val builder = SignatureParameters.builder()
                .component("@method")
                .component("content-type")
                .keyId(KEY_ID)
                .createdEpoch(createdEpoch)
            if (includeAlgorithm) {
                builder.algorithm(Algorithm.HmacSha256)
            }
            val params = builder.build()
            val result = Signer.sign(simpleMessage(), "sig1", params, key)
            return RawMessage.request(
                "POST",
                URI.create("https://example.com/resource"),
                mapOf(
                    "content-type" to listOf("application/json"),
                    "signature-input" to listOf(Signer.signatureInputHeader(result)),
                    "signature" to listOf(Signer.signatureHeader(result)),
                ),
            )
        }
    }

    // ----------------------------------------------------------------
    // 1. Future-dated signature rejection
    // ----------------------------------------------------------------

    @Test
    fun `future-dated signature rejected with tight clock skew`() {
        val now = Instant.ofEpochSecond(1_700_000_000)
        val oneHourAhead = now.plusSeconds(3600).epochSecond

        val signed = signMessage(oneHourAhead)

        val ex = assertThrows<HttpSigException> {
            Verifier.verify(
                signed,
                { _, _ -> hmacKey() },
                Verifier.VerifyOptions(
                    maxClockSkew = Duration.ofSeconds(30),
                    now = Supplier { now },
                ),
            )
        }
        assertNotNull(ex.message)
    }

    @Test
    fun `future-dated signature accepted with generous clock skew`() {
        val now = Instant.ofEpochSecond(1_700_000_000)
        val oneHourAhead = now.plusSeconds(3600).epochSecond

        val signed = signMessage(oneHourAhead)

        val result = Verifier.verify(
            signed,
            { _, _ -> hmacKey() },
            Verifier.VerifyOptions(
                maxClockSkew = Duration.ofHours(2),
                now = Supplier { now },
            ),
        )
        assertEquals("sig1", result.label)
        assertEquals(KEY_ID, result.keyId)
    }

    @Test
    fun `future-dated signature accepted with no clock skew check`() {
        val now = Instant.ofEpochSecond(1_700_000_000)
        val oneHourAhead = now.plusSeconds(3600).epochSecond

        val signed = signMessage(oneHourAhead)

        // null maxClockSkew (the default) means no future-date checking
        val result = Verifier.verify(
            signed,
            { _, _ -> hmacKey() },
            Verifier.VerifyOptions(
                maxClockSkew = null,
                now = Supplier { now },
            ),
        )
        assertEquals("sig1", result.label)
    }

    @Test
    fun `expired signature rejected by default`() {
        val now = Instant.ofEpochSecond(1_700_000_000)
        val expired = now.minusSeconds(60).epochSecond

        val key = hmacKey()
        val params = SignatureParameters.builder()
            .component("@method")
            .keyId(KEY_ID)
            .algorithm(Algorithm.HmacSha256)
            .createdEpoch(now.minusSeconds(120).epochSecond)
            .expiresEpoch(expired)
            .build()
        val result = Signer.sign(simpleMessage(), "sig1", params, key)
        val signed = RawMessage.request(
            "POST",
            URI.create("https://example.com/resource"),
            mapOf(
                "content-type" to listOf("application/json"),
                "signature-input" to listOf(Signer.signatureInputHeader(result)),
                "signature" to listOf(Signer.signatureHeader(result)),
            ),
        )

        assertThrows<HttpSigException> {
            Verifier.verify(
                signed,
                { _, _ -> hmacKey() },
                Verifier.VerifyOptions(now = Supplier { now }),
            )
        }
    }

    @Test
    fun `expired signature accepted when rejectExpired is false`() {
        val now = Instant.ofEpochSecond(1_700_000_000)
        val expired = now.minusSeconds(60).epochSecond

        val key = hmacKey()
        val params = SignatureParameters.builder()
            .component("@method")
            .keyId(KEY_ID)
            .algorithm(Algorithm.HmacSha256)
            .createdEpoch(now.minusSeconds(120).epochSecond)
            .expiresEpoch(expired)
            .build()
        val result = Signer.sign(simpleMessage(), "sig1", params, key)
        val signed = RawMessage.request(
            "POST",
            URI.create("https://example.com/resource"),
            mapOf(
                "content-type" to listOf("application/json"),
                "signature-input" to listOf(Signer.signatureInputHeader(result)),
                "signature" to listOf(Signer.signatureHeader(result)),
            ),
        )

        val verifyResult = Verifier.verify(
            signed,
            { _, _ -> hmacKey() },
            Verifier.VerifyOptions(
                rejectExpired = false,
                now = Supplier { now },
            ),
        )
        assertEquals("sig1", verifyResult.label)
    }

    // ----------------------------------------------------------------
    // 2. Algorithm mismatch rejection
    // ----------------------------------------------------------------

    @Test
    fun `algorithm mismatch between signature-input and key is rejected`() {
        // Sign with HMAC key, including alg="hmac-sha256" in signature params
        val signed = signMessage(1_700_000_000, includeAlgorithm = true)

        // Tamper: rewrite Signature-Input to claim alg="ed25519"
        val origSigInput = signed.headerValues("signature-input").single()
        val tamperedSigInput = origSigInput.replace("alg=\"hmac-sha256\"", "alg=\"ed25519\"")
        // Sanity: make sure the replacement actually changed something
        assert(tamperedSigInput != origSigInput) { "tamper did not modify Signature-Input" }

        val tamperedMsg = RawMessage.request(
            "POST",
            URI.create("https://example.com/resource"),
            mapOf(
                "content-type" to listOf("application/json"),
                "signature-input" to listOf(tamperedSigInput),
                "signature" to signed.headerValues("signature"),
            ),
        )

        // KeyProvider returns the real HMAC key (algorithm = HmacSha256),
        // but the input now claims ed25519, so verification must fail.
        val ex = assertThrows<HttpSigException> {
            Verifier.verify(tamperedMsg, { _, _ -> hmacKey() })
        }
        assertNotNull(ex.message)
    }

    // ----------------------------------------------------------------
    // 3. VerifyResult returns the key's algorithm
    // ----------------------------------------------------------------

    @Test
    fun `verify result contains key algorithm`() {
        val signed = signMessage(1_700_000_000)

        val result = Verifier.verify(signed, { _, _ -> hmacKey() })

        assertEquals(Algorithm.HmacSha256, result.algorithm)
    }
}
